lib/widgets/editor/editor_toolbar.dart

```diff
@@ -0,0 +1,462 @@
+      ),
+    );
+  }
+
+  void reformatSimple(String text) =>
+      reformat((selection) => Reformat(text: text));
+}
+
+enum HeaderLevel {
+  h1(1),
+  h2(2),
+  h3(3),
+  h4(4),
+  h5(5),
+  h6(6);
+
+  const HeaderLevel(this.value);
+  final int value;
+}
+
+class Toolbar extends HookWidget {
+  final TextEditingController controller;
+  final String instanceHost;
+  final EditorToolbarStore store;
+  static const _height = 50.0;
+
+  Toolbar({
+    required this.controller,
+    required this.instanceHost,
+  }) : store = EditorToolbarStore(instanceHost);
+
+  @override
+  Widget build(BuildContext context) {
+    return MobxProvider.value(
+      value: store,
+      child: AsyncStoreListener(
+        asyncStore: store.imageUploadState,
+        child: Container(
+          height: _height,
+          width: double.infinity,
+          color: Theme.of(context).cardColor,
+          child: Material(
+            child: SingleChildScrollView(
+              scrollDirection: Axis.horizontal,
+              child: _ToolbarBody(
+                controller: controller,
+                instanceHost: instanceHost,
+              ),
+            ),
+          ),
+        ),
+      ),
+    );
+  }
+
+  static Widget safeArea = const SizedBox(height: _height);
+}
+
+class _ToolbarBody extends HookWidget {
+  const _ToolbarBody({
+    required this.controller,
+    required this.instanceHost,
+  });
+
+  final TextEditingController controller;
+  final String instanceHost;
+
+  @override
+  Widget build(BuildContext context) {
+    final loggedInAction = useLoggedInAction(instanceHost);
+    return Row(
+      children: [
+        IconButton(
+          onPressed: () => controller.surround(
+              before: '**',
+              placeholder: L10n.of(context).insert_text_here_placeholder),
+          icon: const Icon(Icons.format_bold),
+          tooltip: L10n.of(context).editor_bold,
+        ),
+        IconButton(
+          onPressed: () => controller.surround(
+              before: '*',
+              placeholder: L10n.of(context).insert_text_here_placeholder),
+          icon: const Icon(Icons.format_italic),
+          tooltip: L10n.of(context).editor_italics,
+        ),
+        IconButton(
+          onPressed: () async {
+            final r =
+                await AddLinkDialog.show(context, controller.selectionText);
+            if (r != null) controller.reformat((_) => r);
+          },
+          icon: const Icon(Icons.link),
+          tooltip: L10n.of(context).editor_link,
+        ),
+        // Insert image
+        ObserverBuilder<EditorToolbarStore>(
+          builder: (context, store) {
+            return IconButton(
+              onPressed: loggedInAction((token) async {
+                if (store.imageUploadState.isLoading) {
+                  return;
+                }
+                try {
+                  // FIXME: for some reason it doesn't go past this line on iOS. idk why
```

Comment:
Is that for later?

Reply:
it was a problem that was resolved at very late stage of this PR, can be removed now